### ISSUE TITLE
Freeze after first Heatmap()

### ISSUE THREAD

**Issue by pseudorational** [jokergoo/ComplexHeatmap#782]
Thank you for putting together and maintaining ComplexHeatmap. I just got started with the github version 2.9.3 (July 29, 2021) on R Version 4.0.2. Heatmap() runs without any issue for the first heatmap. But, on trying to repeat this with the same matrix, R freezes. This occurs both with RStudio and from R Console. 

Here are some snippets from running devtools::session_info()
``` r
- Session info -----------------------------------------------------------------------------------
 setting  value                       
 version  R version 4.0.2 (2020-06-22)
 os       Windows 10 x64              
 system   x86_64, mingw32             
 ui       RStudio                     
 language (EN)                        
 collate  English_United States.1252  
 ctype    English_United States.1252  
 tz       America/New_York            
 date     2021-07-29 

 ComplexHeatmap * 2.9.3    2021-07-29 [1] Github (jokergoo/ComplexHeatmap@7e066ad)
```

Thanks,
Vishal

**Comment by jokergoo** [comment]
Can you give me a runnable example?

**Comment by eggrandio** [comment]
> Can you give me a runnable example?

I am no the original poster but I got the same problem. I had used ComplexHeatmap a while ago without any problem at all.
This is the code I run, but it happens with other code that was working perfectly before:

Source data: [https://file.io/BTc5VW2BjU7T](url)

```
mat = as.matrix(read.delim("./input/ForClustering_v3.txt", row.names=1, header=T))

test = Heatmap(mat,
        column_order = colnames(mat),
        clustering_distance_rows = "manhattan",
        clustering_method_rows = "ward.D2",
        na_col = "white",
        show_row_names = T,
        show_column_names = F,
        heatmap_legend_param = list(title = "log2(FC)"),
        row_title = "Genes",
        use_raster = F)
```
When I run it for the first time, I can plot the heatmap, but when I do it for a second time (even without changing the code), R freezes. It keeps running, I can interrupt it, but it does not respond. I have to close it and restart.

```
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19043)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] stringr_1.4.0        ComplexHeatmap_2.9.3

loaded via a namespace (and not attached):
 [1] Biobase_2.52.0         viridis_0.6.1          httr_1.4.2             tidyr_1.1.3            jsonlite_1.7.2        
 [6] bit64_4.0.5            viridisLite_0.4.0      foreach_1.5.1          assertthat_0.2.1       stats4_4.1.0          
[11] blob_1.2.2             GenomeInfoDbData_1.2.6 yaml_2.2.1             pillar_1.6.2           RSQLite_2.2.7         
[16] backports_1.2.1        glue_1.4.2             digest_0.6.27          RColorBrewer_1.1-2     XVector_0.32.0        
[21] colorspace_2.0-2       htmltools_0.5.1.1      pkgconfig_2.0.3        GetoptLong_1.0.5       broom_0.7.9           
[26] zlibbioc_1.38.0        purrr_0.3.4            scales_1.1.1           tibble_3.1.2           KEGGREST_1.32.0       
[31] generics_0.1.0         IRanges_2.26.0         ggplot2_3.3.5          ellipsis_0.3.2         cachem_1.0.5          
[36] BiocGenerics_0.38.0    magrittr_2.0.1         crayon_1.4.1           memoise_2.0.0          evaluate_0.14         
[41] fansi_0.5.0            doParallel_1.0.16      tools_4.1.0            GlobalOptions_0.1.2    lifecycle_1.0.0       
[46] matrixStats_0.60.0     S4Vectors_0.30.0       munsell_0.5.0          cluster_2.1.2          AnnotationDbi_1.54.1  
[51] Biostrings_2.60.1      compiler_4.1.0         GenomeInfoDb_1.28.1    tinytex_0.33           rlang_0.4.11          
[56] RCurl_1.98-1.3         iterators_1.0.13       rstudioapi_0.13        rjson_0.2.20           circlize_0.4.13       
[61] bitops_1.0-7           rmarkdown_2.10         gtable_0.3.0           codetools_0.2-18       DBI_1.1.1             
[66] R6_2.5.0               gridExtra_2.3          knitr_1.33             dplyr_1.0.7            fastmap_1.1.0         
[71] bit_4.0.4              utf8_1.2.1             clue_0.3-59            dendextend_1.15.1      shape_1.4.6           
[76] stringi_1.7.3          parallel_4.1.0         Rcpp_1.0.6             vctrs_0.3.8            png_0.1-7             
[81] tidyselect_1.1.1       xfun_0.24             
```

**Comment by eggrandio** [comment]
> Can you give me a runnable example?

I have tried to run the same code in another computer, where I used ComplexHeatmap before, and I did not get the error. I was able to plot the Heatmap repeatedly. So maybe it's something not related to R? 

In any case, I used an old version of ComplexHeatmap. Here is the sessionInfo.

Please let me know if you need any additional information, I have access to the two computers.

```
R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=Spanish_Spain.1252  LC_CTYPE=Spanish_Spain.1252    LC_MONETARY=Spanish_Spain.1252 LC_NUMERIC=C                  
[5] LC_TIME=Spanish_Spain.1252    

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] stringr_1.4.0        ComplexHeatmap_2.4.3

loaded via a namespace (and not attached):
 [1] shape_1.4.5          circlize_0.4.12      tinytex_0.29         GetoptLong_1.0.5     tidyselect_1.1.0     xfun_0.20           
 [7] purrr_0.3.4          colorspace_2.0-0     vctrs_0.3.6          generics_0.1.0       htmltools_0.5.1.1    stats4_4.0.3        
[13] viridisLite_0.3.0    yaml_2.2.1           blob_1.2.1           rlang_0.4.10         pillar_1.4.7         glue_1.4.2          
[19] DBI_1.1.1            RColorBrewer_1.1-2   BiocGenerics_0.34.0  bit64_4.0.5          lifecycle_1.0.0      munsell_0.5.0       
[25] gtable_0.3.0         GlobalOptions_0.1.2  memoise_2.0.0        evaluate_0.14        Biobase_2.48.0       knitr_1.31          
[31] IRanges_2.22.2       fastmap_1.1.0        parallel_4.0.3       AnnotationDbi_1.50.3 broom_0.7.4          Rcpp_1.0.6          
[37] scales_1.1.1         backports_1.2.1      cachem_1.0.3         S4Vectors_0.26.1     bit_4.0.4            gridExtra_2.3       
[43] rjson_0.2.20         ggplot2_3.3.3        png_0.1-7            digest_0.6.27        stringi_1.5.3        dplyr_1.0.4         
[49] clue_0.3-58          tools_4.0.3          magrittr_2.0.1       tibble_3.0.6         RSQLite_2.2.3        cluster_2.1.0       
[55] crayon_1.4.1         tidyr_1.1.2          pkgconfig_2.0.3      dendextend_1.14.0    ellipsis_0.3.1       assertthat_0.2.1    
[61] rmarkdown_2.6        viridis_0.5.1        R6_2.5.0             compiler_4.0.3      
```

**Comment by jokergoo** [comment]
This is very strange. I ran your code on MacOS, Windows and also Linux, everything ran smoothly. Also the package is tested on the three platforms when I submit a new update on GitHub. Should be of no problem.

Anyway, can you set `ht_opt$verbose = TRUE` before making the heatmap? With the message I might get some clues of what was going on there.

**Comment by eggrandio** [comment]
> This is very strange. I ran your code on MacOS, Windows and also Linux, everything ran smoothly. Also the package is tested on the three platforms when I submit a new update on GitHub. Should be of no problem.
> 
> Anyway, can you set `ht_opt$verbose = TRUE` before making the heatmap? With the message I might get some clues of what was going on there.

This is the output when I run the code for the first time. It runs smoothly and takes less than 5 seconds to generate the Heatmap:
```
> Heatmap(mat,
+         column_order = colnames(mat),
+         clustering_distance_rows = "manhattan",
+         clustering_method_rows = "ward.D2",
+         na_col = "white",
+         show_row_names = T,
+         show_column_names = F,
+         heatmap_legend_param = list(title = "log2(FC)"),
+         row_title = "Genes",
+         use_raster = F)
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
1 heatmaps/annotations, main heatmap: 1th
auto adjust all heatmap/annotations by the main heatmap
apply clustering on each slice (1 slices)
reorder dendrograms in each row slice
perform row clustering on the main heatmap
adjust row order for all other heatmaps
adjust heights for all other heatmaps
remove row titles for all other heatmaps
adjust row_gap for all other heatmaps
prepare layout for heatmap: matrix_1
no clustering is applied/exists on columns
adjust title/dend height of all heatmaps
adjust height to the main heatmap
draw annotation generated by anno_text
```
The second time, it gets stuck at this step:
```
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
```
I hope this helps! Maybe I can try to downgrade to version 2.4.3 in the computer that gives problems if that helps.

**Comment by jokergoo** [comment]
Great! Now there is a direction to look into. Thanks!

**Comment by eggrandio** [comment]
> Great! Now there is a direction to look into. Thanks!

I have tried to run it again without annotations:
```
Heatmap(mat,
        column_order = colnames(mat),
        clustering_distance_rows = "manhattan",
        clustering_method_rows = "ward.D2",
        na_col = "white",
        show_row_names = F,
        show_column_names = F,
        #heatmap_legend_param = list(title = "log2(FC)"),
        #row_title = "Genes",
        use_raster = F)
```
Now it gets stuck at this point (I guess it's the same):
```
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
```

Thanks again for looking into this! This package has been super helpful and I would like to continue using it!

**Comment by jokergoo** [comment]
For this example

```r
Heatmap(mat,
        column_order = colnames(mat),
        clustering_distance_rows = "manhattan",
        clustering_method_rows = "ward.D2",
        na_col = "white",
        show_row_names = F,
        show_column_names = F,
        use_raster = F)
```

Can you run 

```
Heatmap(mat)
```

```
Heatmap(mat,
        column_order = colnames(mat)
)
```

and 

```
Heatmap(mat,
        clustering_distance_rows = "manhattan",
        clustering_method_rows = "ward.D2"
)
```

Do the three all have problems?

Thanks!

**Comment by eggrandio** [comment]
All three had the same problem. It always happens the second time I use Heatmap, even if I am using different datasets and parameters.

```
> Heatmap(mat)
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
construct AnnotationFunction with 'anno_text()'
1 heatmaps/annotations, main heatmap: 1th
auto adjust all heatmap/annotations by the main heatmap
apply clustering on each slice (1 slices)
reorder dendrograms in each row slice
perform row clustering on the main heatmap
adjust row order for all other heatmaps
adjust heights for all other heatmaps
remove row titles for all other heatmaps
adjust row_gap for all other heatmaps
prepare layout for heatmap: matrix_2
apply clustering on each slice (1 slices)
reorder dendrograms in each column slice
adjust title/dend height of all heatmaps
adjust height to the main heatmap
draw annotation generated by anno_text
draw annotation generated by anno_text
> Heatmap(mat)
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
construct AnnotationFunction with 'anno_text()'
```

```
> Heatmap(mat,
+         column_order = colnames(mat)
+ )
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
construct AnnotationFunction with 'anno_text()'
1 heatmaps/annotations, main heatmap: 1th
auto adjust all heatmap/annotations by the main heatmap
apply clustering on each slice (1 slices)
reorder dendrograms in each row slice
perform row clustering on the main heatmap
adjust row order for all other heatmaps
adjust heights for all other heatmaps
remove row titles for all other heatmaps
adjust row_gap for all other heatmaps
prepare layout for heatmap: matrix_1
no clustering is applied/exists on columns
adjust title/dend height of all heatmaps
adjust height to the main heatmap
draw annotation generated by anno_text
draw annotation generated by anno_text
> Heatmap(mat,
+         column_order = colnames(mat)
+ )
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
construct AnnotationFunction with 'anno_text()'
```

```
> Heatmap(mat,
+         clustering_distance_rows = "manhattan",
+         clustering_method_rows = "ward.D2"
+ )
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
construct AnnotationFunction with 'anno_text()'
1 heatmaps/annotations, main heatmap: 1th
auto adjust all heatmap/annotations by the main heatmap
apply clustering on each slice (1 slices)
reorder dendrograms in each row slice
perform row clustering on the main heatmap
adjust row order for all other heatmaps
adjust heights for all other heatmaps
remove row titles for all other heatmaps
adjust row_gap for all other heatmaps
prepare layout for heatmap: matrix_1
apply clustering on each slice (1 slices)
reorder dendrograms in each column slice
adjust title/dend height of all heatmaps
adjust height to the main heatmap
draw annotation generated by anno_text
draw annotation generated by anno_text
> Heatmap(mat,
+         clustering_distance_rows = "manhattan",
+         clustering_method_rows = "ward.D2"
+ )
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
construct AnnotationFunction with 'anno_text()'
construct AnnotationFunction with 'anno_text()'
```

**Comment by jokergoo** [comment]
Can you do one last try?

```r
rownames(mat) = NULL
colnames(mat) = NULL
Heatmap(mat, cluster_rows = FALSE, cluster_columns = FALSE)
```

**Comment by eggrandio** [comment]
This is the output. I wanted to install a previous version of ComplexHeatmaps but I found that I have to reinstall a previous R and Bioconductor versions. Is there an "easy" way of installing previous ComplexHeatmaps versions? (Where can I find the file from previous versions?).

```
> rownames(mat) = NULL
> colnames(mat) = NULL
> Heatmap(mat, cluster_rows = FALSE, cluster_columns = FALSE)
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
1 heatmaps/annotations, main heatmap: 1th
auto adjust all heatmap/annotations by the main heatmap
no clustering is applied/exists on rows
perform row clustering on the main heatmap
adjust row order for all other heatmaps
adjust heights for all other heatmaps
remove row titles for all other heatmaps
adjust row_gap for all other heatmaps
prepare layout for heatmap: matrix_1
no clustering is applied/exists on columns
adjust title/dend height of all heatmaps
adjust height to the main heatmap
> Heatmap(mat, cluster_rows = FALSE, cluster_columns = FALSE)
This matrix has both negative and positive values, use a color mapping symmetric to zero
color is not specified, use randomly generated colors
input color is a color mapping function
```

**Comment by eggrandio** [comment]
Also, it might be something probably related to this specific computer setup.
Running this simple example also gives the same issue:
```
> Heatmap(discrete_mat, name = "mat", col = colors,
+         column_title = "a discrete character matrix")
matrix is character. Do not cluster unless distance method is provided.
input color is a named vector
```

**Comment by eggrandio** [comment]
> Can you do one last try?
> 
> ```r
> rownames(mat) = NULL
> colnames(mat) = NULL
> Heatmap(mat, cluster_rows = FALSE, cluster_columns = FALSE)
> ```

I think I found the issue! It was nothing related to `ComplexHeatmap`. 

I reinstalled all the packagees and found that R was not updating `Rcpp` correctly.

I manually reinstalled the `Rcpp` package it has fixed the problem with `ComplexHeatmap`.

It was not `Rcpp` version as I had 1.0.6 in both computers, but maybe it was somehow "corrupted"?

Thanks again and sorry for making you look into this when it had nothing to do with `ComplexHeatmap` !
